From a4837791f62283079e7be4b17efb769be8b2dfd1 Mon Sep 17 00:00:00 2001
From: Jakub Hrozek <jhrozek@redhat.com>
Date: Thu, 23 Feb 2017 12:22:49 +0100
Subject: [PATCH 93/95] CONFDB: If no configuration file is provided, create a
 fallback configuration
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

This functionality is only enabled in case SSSD is configured with with
--enable-files-domain. If not, the behaviour is as it used to -- SSSD
returns an error, instructing the admin to create a configuration file.

If the option is enabled, a very minimal confdb that only enables the
NSS responder is created. The confdb later adds the implicit files
domain.

Resolves:
https://pagure.io/SSSD/sssd/issue/2229

Reviewed-by: Pavel Březina <pbrezina@redhat.com>
---
 src/confdb/confdb_setup.c             | 60 +++++++++++++++++++++++++++--------
 src/tests/intg/test_files_provider.py | 16 ++++++++++
 2 files changed, 63 insertions(+), 13 deletions(-)

diff --git a/src/confdb/confdb_setup.c b/src/confdb/confdb_setup.c
index c040bdf9e59f47541f2d598a37699d035eb9e566..7884eea6367390c41f64b7023fe505dd653b49db 100644
--- a/src/confdb/confdb_setup.c
+++ b/src/confdb/confdb_setup.c
@@ -28,6 +28,14 @@
 #include "confdb_setup.h"
 #include "util/sss_ini.h"
 
+#ifndef SSSD_FALLBACK_CONFIG_LDIF
+#define SSSD_FALLBACK_CONFIG_LDIF \
+"dn: cn=config\n" \
+"version: 2\n\n" \
+"dn: cn=sssd,cn=config\n" \
+"cn: sssd\n" \
+"services: nss\n\n"
+#endif /* SSSD_FALLBACK_CONFIG_LDIF */
 
 static int confdb_test(struct confdb_ctx *cdb)
 {
@@ -227,6 +235,23 @@ static int confdb_ldif_from_ini_file(TALLOC_CTX *mem_ctx,
     return EOK;
 }
 
+static int confdb_fallback_ldif(TALLOC_CTX *mem_ctx,
+                                const char **_timestr,
+                                const char **_ldif)
+{
+#ifndef ADD_FILES_DOMAIN
+    return ERR_MISSING_CONF;
+#else
+    *_timestr = talloc_strdup(mem_ctx, "1");
+    *_ldif = talloc_strdup(mem_ctx, SSSD_FALLBACK_CONFIG_LDIF);
+    if (*_timestr == NULL || *_ldif == NULL) {
+        return ENOMEM;
+    }
+
+    return EOK;
+#endif
+}
+
 static int confdb_init_db(const char *config_file, const char *config_dir,
                           struct confdb_ctx *cdb)
 {
@@ -255,25 +280,34 @@ static int confdb_init_db(const char *config_file, const char *config_dir,
 
     /* Open config file */
     ret = sss_ini_config_file_open(init_data, config_file);
-    if (ret != EOK) {
+    if (ret == EOK) {
+        ret = confdb_ldif_from_ini_file(tmp_ctx,
+                                        config_file,
+                                        config_dir,
+                                        init_data,
+                                        &timestr,
+                                        &config_ldif);
+        if (ret != EOK) {
+            DEBUG(SSSDBG_CRIT_FAILURE,
+                  "Cannot convert INI to LDIF [%d]: [%s]\n",
+                  ret, sss_strerror(ret));
+            goto done;
+        }
+    } else if (ret == ENOENT) {
+        ret = confdb_fallback_ldif(tmp_ctx, &timestr, &config_ldif);
+        if (ret != EOK) {
+            DEBUG(SSSDBG_CRIT_FAILURE,
+                  "Cannot create a fallback configuration [%d]: [%s]\n",
+                  ret, sss_strerror(ret));
+            goto done;
+        }
+    } else {
         DEBUG(SSSDBG_CONF_SETTINGS,
               "sss_ini_config_file_open failed: %s [%d]\n", sss_strerror(ret),
               ret);
         goto done;
     }
 
-    ret = confdb_ldif_from_ini_file(tmp_ctx,
-                                    config_file,
-                                    config_dir,
-                                    init_data,
-                                    &timestr,
-                                    &config_ldif);
-    if (ret != EOK) {
-        DEBUG(SSSDBG_CRIT_FAILURE,
-              "Cannot convert INI to LDIF [%d]: [%s]\n",
-              ret, sss_strerror(ret));
-        goto done;
-    }
     DEBUG(SSSDBG_CONF_SETTINGS, "LDIF file to import: \n%s\n", config_ldif);
 
     /* Set up a transaction to replace the configuration */
diff --git a/src/tests/intg/test_files_provider.py b/src/tests/intg/test_files_provider.py
index abf836bcc7a1810ea505dcf84ad1edfadf0c382c..8748ac10b089087056b1b93950c8d890a190c8d0 100644
--- a/src/tests/intg/test_files_provider.py
+++ b/src/tests/intg/test_files_provider.py
@@ -167,6 +167,12 @@ def disabled_files_domain(request):
     return None
 
 
+@pytest.fixture
+def no_sssd_conf(request):
+    create_sssd_fixture(request)
+    return None
+
+
 def setup_pw_with_list(request, user_list):
     pwd_ops = passwd_ops_setup(request)
     for user in user_list:
@@ -832,3 +838,13 @@ def test_disable_files_domain(add_user_with_canary, disabled_files_domain):
     # The local user will not be resolvable through nss_sss now
     res, user = sssd_getpwnam_sync(USER1["name"])
     assert res != NssReturnCode.SUCCESS
+
+
+def test_no_sssd_conf(add_user_with_canary, no_sssd_conf):
+    """
+    Test that running without sssd.conf implicitly configures one with
+    id_provider=files
+    """
+    res, user = sssd_getpwnam_sync(USER1["name"])
+    assert res == NssReturnCode.SUCCESS
+    assert user == USER1
-- 
2.9.3

